src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs

```diff
@@ -36,13 +38,55 @@ internal abstract partial class AbstractDiagnosticsTaggerProvider<TTag> : Asynch
     {
         private readonly IDiagnosticService _diagnosticService;
 
+        /// <summary>
+        /// Keep track of the ITextSnapshot for the open Document that was used when diagnostics were
+        /// produced for it.  We need that because the DiagnoticService does not keep track of this
+        /// snapshot (so as to not hold onto a lot of memory), which means when we query it for 
+        /// diagnostics, we don't know how to map the span of the diagnostic to the current snapshot
+        /// we're tagging.
+        /// </summary>
+        private static readonly ConditionalWeakTable<object, ITextSnapshot> _diagnosticIdToTextSnapshot = 
+            new ConditionalWeakTable<object, ITextSnapshot>();
```

Comment:
@heejaechang I think the lifetimes here are exactly what we want.  Namely, as long as the diagnostics subsystem is holding onto these IDs, then we'll hold onto the snapshot here.  And the diagnostic subsystem holds onto this ID until it is replaced/removed with the most up to date version of the diagnostics for that analyzer/document pair, right?

Reply:
basically, as long as we could observe the diagnostics associated with an id, then we'll have the snapshot for it.  And once we can't observe them anymore (because they were replaced/removed), and thus GetDiagnosticsUpdatedEventArgs won't return them, then this will release the snapshot as well.

Reply:
yep.